### ISSUE TITLE
Enum support

### ISSUE THREAD

**Issue by franciscolourenco** [khaosdoctor/gotql#36]
**Describe the bug**
[Hasura's `on_conflict` argument](https://hasura.io/docs/1.0/graphql/manual/mutations/upsert.html#update-selected-columns-on-conflict) expect enums to be provided without quotes, however gotql does use quotes for every argument value. This result in an error (see screenshots).

**To Reproduce**
```js
gotQl.mutation(
  process.env.GRAPHQL_HOST,
  {
    operation: {
      name: 'insert_example',
      args: {
        objects: {id, status, price},
        on_conflict: {constraint: 'example_pkey', update_columns: ['status', 'price']},
      },
      fields: ['affected_rows'],
    }
  },
)
```
**Current behavior**
Part of the query:
```
on_conflict: {constraint: \"example_pkey\", update_columns: [\"status\", \"price\"]}
```

**Expected behavior**
Part of the query:
```
on_conflict: {constraint: example_pkey, update_columns: [status, price]}
```

**Screenshots**
![image](https://user-images.githubusercontent.com/208149/81201520-a81ca400-8fc5-11ea-95f6-907f6f5d2013.png)
![image](https://user-images.githubusercontent.com/208149/81201599-bec2fb00-8fc5-11ea-81e9-381e8de49e3f.png)
![image](https://user-images.githubusercontent.com/208149/81201679-d4382500-8fc5-11ea-8990-01ab6c1b9d15.png)

**Comment by franciscolourenco** [comment]
I guess I found my answer: https://github.com/khaosdoctor/gotql#enum-args

**Comment by franciscolourenco** [comment]
Re-opening because this:
```js
on_conflict: {
  constraint: {value: 'subscriptions_pkey', escape: false},
  update_columns: [
    {value: 'status', escape: false},
    {value: 'price', escape: false},
  ],
},
```
results in this query:
```
on_conflict: { constraint: subscriptions_pkey, update_columns: [\"[object Object]\",\"[object Object]\"] })
```

**Comment by franciscolourenco** [comment]
```js
on_conflict: {
  constraint: {value: 'subscriptions_pkey', escape: false},
  update_columns: {
    value: ['status', 'price'],
    escape: false,
  },
},
```
almost works:
```
on_conflict: { constraint: subscriptions_pkey, update_columns: status,price }
```

**Comment by khaosdoctor** [comment]
Escaped values will not be processed, for you to obtain the correct representation, you need to transform your query into:

```js
on_conflict: {
  constraint: {value: 'subscriptions_pkey', escape: false},
  update_columns: {
    value: '[status,price]',
    escape: false,
  },
},
```

**Comment by franciscolourenco** [comment]
Thanks. It might be worth to enrich the docs with this example. This edge case makes the concept of this library a bit more verbose than expected. Have you thought of providing a helper function for escaped values? Example:

```js
import {literal} from 'gotql'

on_conflict: {
  constraint: literal('subscriptions_pkey'),
  update_columns: literal(['status','price']),
},
```

**Comment by franciscolourenco** [comment]
or with a tagged template literal
```js
import {literal} from 'gotql'

on_conflict: {
  constraint: literal`subscriptions_pkey`,
  update_columns: [literal`status`, literal`price`],
},
```

**Comment by khaosdoctor** [comment]
> Thanks. It might be worth to enrich the docs with this example. 

Yeah, I'll update the README with some examples like these to make it clearer.

> This edge case makes the concept of this library a bit more verbose than expected. Have you thought of providing a helper function for escaped values? 

I thought about it recently, when I created it there weren't this many resources on ECMA yet. However, the whole idea of gotql is not to make queries smaller, but clearer to understand and to manipulate. Even with simple queries, the JSON-Like format is more verbose than the common syntax, however, it is way more processable and usable than strings. 

But that might be a good solution to escaped values, I think I'll add it as a function but keeping the support for `{value: x, escape: b}` for now, then I'll release a V2 with no support to those at all. 

> or with a tagged template literal
> 
> ```js
> import {literal} from 'gotql'
> 
> on_conflict: {
>   constraint: literal`subscriptions_pkey`,
>   update_columns: [literal`status`, literal`price`],
> },
> ```

This is cleaner, but unfortunately, it isn't possible. The parser parses the object as a whole, so when you put ```[literal`status`]``` the brackets `[]` identify an array and the parser will identify the outside object as an array regardless of what comes inside, it'd be a big change in the way it works.

For now, the best course of action is to work as I said before, but I'll implement a `literal` function along with some refactoring (to move libs and helpers outside some files) which converts whatever is inside to `{value: x, escape: b}`.

Thanks for the heads up and the solution

**Comment by khaosdoctor** [comment]
> Thanks. It might be worth to enrich the docs with this example. 

Yeah, I'll update the README with some examples like these to make it clearer.

> This edge case makes the concept of this library a bit more verbose than expected. Have you thought of providing a helper function for escaped values? 

I thought about it recently, when I created it there weren't this many resources on ECMA yet. However, the whole idea of gotql is not to make queries smaller, but clearer to understand and to manipulate. Even with simple queries, the JSON-Like format is more verbose than the common syntax, however, it is way more processable and usable than strings. 

But that might be a good solution to escaped values, I think I'll add it as a function but keeping the support for `{value: x, escape: b}` for now, then I'll release a V2 with no support to those at all. 

> or with a tagged template literal
> 
> ```js
> import {literal} from 'gotql'
> 
> on_conflict: {
>   constraint: literal`subscriptions_pkey`,
>   update_columns: [literal`status`, literal`price`],
> },
> ```

This is cleaner, but unfortunately, it isn't possible. The parser parses the object as a whole, so when you put ```[literal`status`]``` the brackets `[]` identify an array and the parser will identify the outside object as an array regardless of what comes inside, it'd be a big change in the way it works.

For now, the best course of action is to work as I said before, but I'll implement a `literal` function along with some refactoring (to move libs and helpers outside some files) which converts whatever is inside to `{value: x, escape: b}`.

Thanks for the heads up and the solution

**Comment by khaosdoctor** [comment]
Actually, thanks for using the lib and providing such useful feedback and bug reports in general, you're evolving it so other people can benefit from it as well. Since I'm not actively using gotQL a lot, this is very helpful to those who will use in the future and allow the community to grow and the lib to be even better.

Thanks a lot!

**Comment by franciscolourenco** [comment]
Thank you for creating and maintaining :) 
All the helper function needs to do really is return an object `{value: x, escape: b}`

**Comment by khaosdoctor** [comment]
Yep, just what I was thinking :D

**Comment by khaosdoctor** [comment]
Closed in #47
